### PR TITLE
[SPARK-8029] Robust shuffle writer

### PR DESCRIPTION
**Pull request by davies**
Currently, all the shuffle writer will write to target path directly, the file could be corrupted by other attempt of the same partition on the same executor. They should write to temporary file then rename to target path, as what we do in output committer. In order to make the rename atomic, the temporary file should be created in the same local directory (FileSystem).

This PR is based on #9214 , thanks to @squito . Closes #9214

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 1]
```diff
@@ -93,6 +95,10 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+    indexFile.deleteOnExit()
+    if (!tmp.renameTo(indexFile)) {
+      throw new IOException(s"fail to rename index file $tmp to $indexFile")
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: this will just kill the task, right?  both tasks are actually just fine, and in fact the overall job should continue if one of them succeeds.  But instead this will lead to the task getting retried, and potentially continuing to fail up to 4 times, though its actually finished successfully from another taskset?  You could handle this in scheduler, but that would add some complexity.

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java [hunk 2]
```diff
@@ -155,9 +156,20 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
       writer.commitAndClose();
     }
 
-    partitionLengths =
-      writePartitionedFile(shuffleBlockResolver.getDataFile(shuffleId, mapId));
-    shuffleBlockResolver.writeIndexFile(shuffleId, mapId, partitionLengths);
+    File output = shuffleBlockResolver.getDataFile(shuffleId, mapId);
+    final File tmp = new File(output.getAbsolutePath() + "." + UUID.randomUUID());
+    partitionLengths = writePartitionedFile(tmp);
+    if (!output.exists()) {
```
Comment (squito) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: I dont' think you can do this and still support SPARK-4085 -- regenerating the output if one of the shuffle files goes completely missing.  Because if the index file goes missing, and the data file is still there, with this logic you'll always never regenerate the shuffle output.  But maybe SPARK-4085 is not worth it ...

File: core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala [hunk 3]
```diff
@@ -106,6 +108,19 @@ private[spark] class HashShuffleWriter[K, V](
...
     }
+    // rename all shuffle files to final paths
+    shuffle.writers.zip(sizes).foreach { case (writer: DiskBlockObjectWriter, size: Long) =>
+      if (size > 0) {
+        val output = blockManager.diskBlockManager.getFile(writer.blockId)
+        if (output.exists()) {
+          writer.file.delete()
+        } else {
+          if (!writer.file.renameTo(output)) {
+            throw new IOException(s"fail to rename ${writer.file} to $output")
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala]: same problem here on partially existing shuffle output.  Also, the `if (size > 0)` check will lead to inconsistencies if you have non-deterministic shuffle output -- it might be true for different partitions in different attempts.  I think it can be OK in any case, as long as the MapStatus is consistent.

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java [hunk 4]
```diff
@@ -155,9 +156,20 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
       writer.commitAndClose();
     }
 
-    partitionLengths =
-      writePartitionedFile(shuffleBlockResolver.getDataFile(shuffleId, mapId));
-    shuffleBlockResolver.writeIndexFile(shuffleId, mapId, partitionLengths);
+    File output = shuffleBlockResolver.getDataFile(shuffleId, mapId);
+    final File tmp = new File(output.getAbsolutePath() + "." + UUID.randomUUID());
```
Comment (squito) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: good point about creating the tmp files in the same dir as the dest to make sure we can do the rename ... I had taken that for granted.

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java (hunk 2 above)
Comment (davies) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: Good point, we should check both.

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala (hunk 1 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: There is very little chance that the two concurrent task will call `renameTo` in the same time, even with that, one of them will succeed, the scheduler will mark the partition as success, and the failure will be ignored (not retried).

File: core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala (hunk 3 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala]: I think these partitions are independent, they should be OK whenever it's generated in different attempt, or that's the basic idea of how RDD works (could be re-run and got the same result).

Comment (davies) [inline core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala]: The `size` here is used to make sure that a new file is generated (we could try to delete existed file if size is 0).

**Comment by davies** [comment]
cc @mateiz @rxin

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala (hunk 1 above)
Comment (mateiz) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: Can you test for this? I think the worry was about different TaskSets attempting the same map stage. Imagine that attempt 1 of the stage successfully completes a task, and sends back a map output status, but that status gets ignored because that stage attempt got cancelled. Attempt 2 might then fail to send a new status for it.

There seem to be two ways to fix it if this problem can actually occur -- either add MapOutputStatuses even from failed task sets or mark this new task as successful if a file exists.

Comment (davies) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: On Thu, Nov 12, 2015 at 8:30 AM, Matei Zaharia <notifications@github.com>
wrote:

> In
> core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala
> <https://github.com/apache/spark/pull/9610#discussion_r44678796>:
>
> > @@ -93,6 +95,10 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
> >      } {
> >        out.close()
> >      }
> > +    indexFile.deleteOnExit()
> > +    if (!tmp.renameTo(indexFile)) {
> > +      throw new IOException(s"fail to rename index file $tmp to $indexFile")
>
> Can you test for this? I think the worry was about different TaskSets
> attempting the same map stage. Imagine that attempt 1 of the stage
> successfully completes a task, and sends back a map output status, but that
> status gets ignored because that stage attempt got cancelled. Attempt 2
> might then fail to send a new status for it.
>
> There seem to be two ways to fix it if this problem can actually occur --
> either add MapOutputStatuses even from failed task sets or mark this new
> task as successful if a file exists.
>
After this PR, the second attempt of same task will return SUCCESS, with
new MapOutputStatus, which could be different than the previous attempt
(having different sizes of partitions), since we does not use the exact
number of size (could be lossy compressed), I think it's fine.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/9610/files#r44678796>.
>



-- 
 - Davies

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 5]
```diff
@@ -93,6 +93,29 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+          indexFile.delete()
+        }
+        if (!indexTmp.renameTo(indexFile)) {
+          throw new IOException("fail to rename data file " + indexTmp + " to " + indexFile)
+        }
+        if (dataFile.exists()) {
+          dataFile.delete()
+        }
+        if (dataTmp != null && dataTmp.exists() && !dataTmp.renameTo(dataFile)) {
+          throw new IOException("fail to rename data file " + dataTmp + " to " + dataFile)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: I don't think there is a particular flaw here, but its a bit hard to follow since its a mix of first-attempt-wins and last-attempt wins.  First attempt if there is a data file & index file; last attempt if its only an index file.  the problem w/ last-attempt is that this delete will fail on windows if the file is open for reading, I believe.  Though we can't get around that because SPARK-4085 always requires us to delete some files that might be open, in which case we hope that we don't run into this race again on the next retry.  It would be nice to minimize that case, though.  We'd be closer to first-attempt-wins if we always wrote a dataFile, even if its empty when dataTmp == null.

There is also an issue w/ mapStatus & non-deterministic data.  It might not matter which output you get, but the mapstatus should be consistent with the data that is read.  If attempt 1 writes non-empty outputs a,b,c, and attempt 2 writes non-empty outputs d,e,f (which are not committed), the reduce tasks might get the mapstatus for attempt 2, look for outputs d,e,f, and get nothing but empty blocks.  Matei had suggested writing the mapstatus to a file, so that subsequent attempts always return the map status corresponding to the first successful attempt.

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 6]
```diff
@@ -93,6 +93,29 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    synchronized {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: Can you add a big comment explaining what is going on here?  Also worth noting that there is only one `IndexShuffleBlockResolver` per executor.

File: core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala (hunk 3 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala]: yeah I suppose it all depends on what the model is for non-deterministic data.  The reduce tasks can read data from a mix attempts, but I guess that is OK (we can't completely prevent it in any case).  There is also the problem of returning the right mapstatus here, but it doesn't matter as much in this case -- you will at least return some set of non-empty blocks that is consistent with the shuffle data on disk, even if the sizes can be arbitrarily wrong.

Also I know its super-rare, but there _is_ a race between `output.exists` and `renameTo(output)`, might as well protect against that.

I also find it a weird that this is neither first or last attempt wins -- the first attempt to get to each output file wins, but it can be a mix of attempts.  again I'd include a comment explaining the logic

**Comment by squito** [comment]
you can take this test case if you like: https://github.com/squito/spark/blob/SPARK-8029_first_wins/core/src/test/scala/org/apache/spark/ShuffleSuite.scala#L351

I'd also add more test cases to cover the various paths through the output commit code.  And I think that `ShuffleWriter.write` should at least include a comment explaining the need for shuffle writers to write to a tmp file and atomically move them to the final place, since its not obvious why that would be necessary, and I guess there isn't any other good place to stick the explanation.

**Comment by davies** [comment]
@squito Thanks for reviewing this, I had included your regression test here, also added tests for resolver.

File: core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java [hunk 7]
```diff
@@ -248,8 +250,7 @@ void forceSorterToSpill() throws IOException {
    *
    * @return the partition lengths in the merged file.
    */
-  private long[] mergeSpills(SpillInfo[] spills) throws IOException {
-    final File outputFile = shuffleBlockResolver.getDataFile(shuffleId, mapId);
+  private long[] mergeSpills(SpillInfo[] spills, File outputFile) throws IOException {
```
Comment (andrewor14) [inline core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java]: This is only used in 1 place so we don't need to change the method signature here. We can just do the `Utils.tempFileWith` in this method.

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 8]
```diff
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    // Note: there is only one IndexShuffleBlockResolver per executor
+    synchronized {
+      val existedLengths = checkIndexAndDataFile(indexFile, dataFile, lengths.length)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: `existingLengths`

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 9]
```diff
@@ -75,13 +81,68 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
   }
 
   /**
+   * Check whether there are index file and data file also they are matched with each other, returns
+   * the lengths of each block in data file, if there are matched, or return null.
+   */
+  private def checkIndexAndDataFile(index: File, data: File, blocks: Int): Array[Long] = {
+    val lengths = new Array[Long](blocks)
+    if (index.length() == (blocks + 1) * 8) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: instead of the nested `if-else`'s, it might be clearer to write this as:
```
// can you add a comment here to explain this check
if (index.length() != (blocks + 1) * 8) {
  return null
}
```

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 10]
```diff
@@ -75,13 +81,68 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+      // Read the lengths of blocks
+      val f = try {
+        new FileInputStream(index)
+      } catch {
+        case e: IOException =>
+          return null
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: should we at least `logWarning` when this happens? Same in L113

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 11]
```diff
@@ -40,10 +39,17 @@ import IndexShuffleBlockResolver.NOOP_REDUCE_ID
...
+    _blockManager: BlockManager = null)
+  extends ShuffleBlockResolver
   with Logging {
 
-  private lazy val blockManager = SparkEnv.get.blockManager
+  private lazy val blockManager = if (_blockManager == null) {
+    SparkEnv.get.blockManager
+  } else {
+    _blockManager
+  }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: or
```
Option(_blockManager).getOrElse(SparkEnv.getBlockManager)
```

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 12]
```diff
@@ -75,13 +81,68 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+      } catch {
+        case e: IOException =>
+          return null
+      }
+      val in = new DataInputStream(new BufferedInputStream(f))
+      try {
+        // Convert the offsets into lengths of each block
+        var offset = in.readLong()
+        if (offset != 0L) {
+          return null
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]:
```
// first offset should be 0
return null
```

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 13]
```diff
@@ -75,13 +81,68 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+      } catch {
+        case e: IOException =>
+          return null
+      } finally {
+        in.close()
+      }
+
+      val size = lengths.reduce(_ + _)
+      // `length` returns 0 if it not exists.
+      if (data.length() == size) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: or just `if (data.length() == lengths.sum)`

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 14]
```diff
@@ -75,13 +81,68 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
   }
 
   /**
+   * Check whether there are index file and data file also they are matched with each other, returns
+   * the lengths of each block in data file, if there are matched, or return null.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]:
```
/**
 * Check whether the given index and data files match each other.
 * If so, return the length of each block in the data file, otherwise return null.
 */
```

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 15]
```diff
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+        indexTmp.delete()
+      } else {
+        if (indexFile.exists()) {
+          indexFile.delete()
+        }
+        if (dataFile.exists()) {
+          dataFile.delete()
+        }
+        if (!indexTmp.renameTo(indexFile)) {
+          throw new IOException("fail to rename data file " + indexTmp + " to " + indexFile)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: data -> index

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 16]
```diff
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
     }
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    // Note: there is only one IndexShuffleBlockResolver per executor
+    synchronized {
+      val existedLengths = checkIndexAndDataFile(indexFile, dataFile, lengths.length)
+      if (existedLengths != null) {
+        // Use the lengths of existed output for MapStatus
+        System.arraycopy(existedLengths, 0, lengths, 0, lengths.length)
+        dataTmp.delete()
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: will this throw NPE?

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 17]
```diff
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    // Note: there is only one IndexShuffleBlockResolver per executor
+    synchronized {
+      val existedLengths = checkIndexAndDataFile(indexFile, dataFile, lengths.length)
+      if (existedLengths != null) {
+        // Use the lengths of existed output for MapStatus
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]:
```
// A previous attempt has already written the map outputs for this task,
// so just use them and delete the ones we've written
```

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 18]
```diff
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+        System.arraycopy(existedLengths, 0, lengths, 0, lengths.length)
+        dataTmp.delete()
+        indexTmp.delete()
+      } else {
+        if (indexFile.exists()) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]:
```
// This is the first successful attempt in writing the map outputs for this task,
// so override any existing index and data files with the ones we wrote.
```

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 19]
```diff
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+        }
+        if (dataFile.exists()) {
+          dataFile.delete()
+        }
+        if (!indexTmp.renameTo(indexFile)) {
+          throw new IOException("fail to rename data file " + indexTmp + " to " + indexFile)
+        }
+        if (dataTmp != null && dataTmp.exists() && !dataTmp.renameTo(dataFile)) {
+          throw new IOException("fail to rename data file " + dataTmp + " to " + dataFile)
+        }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: should we check
```
checkIndexAndDataFile(indexTmp, dataTmp, lengths.length) != null
```
what happens if this fails? Should we throw an exception?

File: core/src/main/scala/org/apache/spark/shuffle/FileShuffleBlockResolver.scala [hunk 20]
```diff
@@ -84,17 +86,8 @@ private[spark] class FileShuffleBlockResolver(conf: SparkConf)
...
-          if (blockFile.exists) {
-            if (blockFile.delete()) {
-              logInfo(s"Removed existing shuffle file $blockFile")
-            } else {
-              logWarning(s"Failed to remove existing shuffle file $blockFile")
-            }
-          }
-          blockManager.getDiskWriter(blockId, blockFile, serializerInstance, bufferSize,
-            writeMetrics)
+          val tmp = new File(blockFile.getAbsolutePath + "." + UUID.randomUUID())
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/FileShuffleBlockResolver.scala]: why not use your new method `Utils.withTempFile`

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java [hunk 21]
```diff
@@ -155,9 +155,10 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
       writer.commitAndClose();
     }
 
-    partitionLengths =
-      writePartitionedFile(shuffleBlockResolver.getDataFile(shuffleId, mapId));
-    shuffleBlockResolver.writeIndexFile(shuffleId, mapId, partitionLengths);
+    File output = shuffleBlockResolver.getDataFile(shuffleId, mapId);
```
Comment (andrewor14) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: `// Write to a temporary location to avoid potential corruption issues (SPARK-8029)`

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala [hunk 22]
```diff
@@ -65,11 +66,11 @@ private[spark] class SortShuffleWriter[K, V, C](
     // Don't bother including the time to open the merged output file in the shuffle write time,
     // because it just opens a single file, so is typically too fast to measure accurately
     // (see SPARK-3570).
-    val outputFile = shuffleBlockResolver.getDataFile(dep.shuffleId, mapId)
+    val output = shuffleBlockResolver.getDataFile(dep.shuffleId, mapId)
+    val tmp = Utils.tempFileWith(output)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala]: can you call these `outputTmp` or something so it's slightly easier to follow? (here and other places)

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 23]
```diff
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    // Note: there is only one IndexShuffleBlockResolver per executor
+    synchronized {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: is this synchronized for atomic renames? If so, can you mention in a comment?

File: core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala [hunk 24]
```diff
@@ -106,6 +108,28 @@ private[spark] class HashShuffleWriter[K, V](
...
       writer.fileSegment().length
     }
+    // rename all shuffle files to final paths
+    // Note: there is only one ShuffleBlockResolver in executor
+    shuffleBlockResolver.synchronized {
+      shuffle.writers.zipWithIndex.foreach { case (writer, i) =>
+        val output = blockManager.diskBlockManager.getFile(writer.blockId)
+        if (sizes(i) > 0) {
+          if (output.exists()) {
+            // update the size of output for MapStatus
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala]: or
```
// Use length of existing file and delete our own temporary one
```

File: core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala [hunk 25]
```diff
@@ -106,6 +108,28 @@ private[spark] class HashShuffleWriter[K, V](
...
+            sizes(i) = output.length()
+            writer.file.delete()
+          } else {
+            if (!writer.file.renameTo(output)) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala]: `// Commit by renaming our temporary file to something the shuffle reader expects`

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala [hunk 26]
```diff
@@ -75,13 +81,68 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
...
+    }
+  }
+
+  /**
    * Write an index file with the offsets of each block, plus a final offset at the end for the
    * end of the output file. This will be used by getBlockData to figure out where each block
    * begins and ends.
+   *
+   * It will commit the data and index file as an atomic operation, use the existed ones (lengths of
+   * blocks will be refreshed), or replace them with new ones.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: we should add that this modifies the contents of `lengths` if existing data and index files exist and are matching.

**Comment by andrewor14** [comment]
@davies I took a pass and I find the approach correct and simple. I did a close review and confirmed that all four `ShuffleWriter`s write temporary files and rename them correctly on commit. Most of my comments are minor.

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala (hunk 19 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala]: This will slowdown the normal path, I think it's not needed.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala (hunk 22 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala]: This is only one file `output` here, I think it's obvious

**Comment by davies** [comment]
@andrewor14 Thanks, this looks much better now.

**Comment by davies** [comment]
The failed test is not related, I'm merging this into master, will create another PR for other branches.

**Comment by Prasidhdh** [comment]
@davies :  If we want to merge two datafiles (one from first action of rdd and another from second action of same rdd), How can i do that? And do i need to do anything with indexfile?
Please give help me to understand this problem!
Thank you
